query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/PostgresClientFactory.java

```diff
@@ -58,32 +81,56 @@ public PostgresClient getPostgresClient(String clusterName) {
 
   public static class PostgresClient {
 
-    private final Connection connection;
+    private final DataSource dataSource;
 
     private PostgresClient(RequestHandlerClientConfig clientConfig) throws SQLException {
       String user = clientConfig.getUser().orElseThrow(IllegalArgumentException::new);
       String password = clientConfig.getPassword().orElseThrow(IllegalArgumentException::new);
-      String url =
-          String.format(
-              "%s?user=%s&password=%s", clientConfig.getConnectionString(), user, password);
-      LOG.debug(
-          "Trying to create a Postgres client connected to postgres server using url: {}", url);
-      this.connection = DriverManager.getConnection(url);
+      String url = clientConfig.getConnectionString();
+      this.dataSource = createPooledDataSource(url, user, password);
     }
 
-    private PostgresClient(Connection connection) {
-      this.connection = connection;
+    private DataSource createPooledDataSource(String url, String user, String password) {
```

Comment:
This is getting really messy really quickly. Why do we need this here and not in doc store for example? Should the postgres client building code be in doc store and shared here?

Reply:
Document store and query service are different.
Document store handles query for data stored within a json document.
Query service handles query for data stored in multiple column within a table as of now.
In short term, they will be separate impl.

Reply:
Agreed, but just to clarify I meant doc store the repo rather than doc store the lib. Both the doc store lib and query service talk to postgres, so using a common postgres connection/client lib between them would make sense, no? (this is the bit that I was suggesting could be stored in doc store the repo)

Reply:
Will move to postgres repo and use it from both places. However it will be done as a followup.